ZCRMiOS/Classes/DashBoardAPIHandler.swift

```diff
@@ -0,0 +1,1737 @@
+        
+        if let label = dict[.label] as? String,
+            let type = dict[.type] as? String,
+            let name = dict[.name] as? String,
+            let aggregation = dict[.aggregation] as? [String] {
+            
+            return AggregateColumn(label: label,
+                                   type: type,
+                                   name: name,
+                                   decimalPlaces: decimalPlaces,
+                                   aggregation: aggregation)
+            
+        }
+        
+        print(debugMsg)
+        return nil
+        
+    }
+    
+    //MARK:- GROUPING COLUMN INFO
+    
+    internal func getArrayOfGroupingColumnInfo(Using componentChunksJSON: [String:Any]) -> [GroupingColumn] {
+        
+        var groupingColumnValues = [GroupingColumnKeys:Any]()
+        
+        var ArrayOfGroupingColumnObj = [GroupingColumn]()
+        
+        let ArrayOfGroupingColumnJSON = componentChunksJSON.getArrayOfDictionaries(key: componentAPINames.groupingColumn)
+        
+        for groupingColumnJSON in ArrayOfGroupingColumnJSON {
+            
+            for (key,value) in groupingColumnJSON {
+                
+                switch (key) {
+                    
+                case componentAPINames.label:
+                    groupingColumnValues[.label] = value
+                    
+                case componentAPINames.type:
+                    groupingColumnValues[.type] = value
+                    
+                case componentAPINames.name:
+                    groupingColumnValues[.name] = value
+                    
+                case componentAPINames.groupingConfig:
+                    let groupingConfigTuple = getGroupingConfigValuesFrom(groupingColumnJSON)
+                    groupingColumnValues[.allowedValues] = groupingConfigTuple.AllowedValues
+                    groupingColumnValues[.customGroups] = groupingConfigTuple.CustomGroups
+                    
+                default:
+                    print("UNKNOWN KEY \(key) ENCOUNTERED IN GROUPING COLUMN PARSING")
+                }
+                
+            } // inner loop
+            
+            if let groupingColumnObj = constructGroupingColumnObjFrom(groupingColumnValues) {
+                ArrayOfGroupingColumnObj.append(groupingColumnObj)
+            }
+            
+        } // outer loop
+        
+        return ArrayOfGroupingColumnObj
+        
+    } // func ends
+    
+    
+    
+    internal func constructGroupingColumnObjFrom(_ dict:[GroupingColumnKeys:Any])
+        -> GroupingColumn? {
+            
+            
+            
+            // Both are optional values
+            let allowedValues = dict[.allowedValues] as? [AllowedValues]
+            let customGroups = dict[.customGroups] as? [String]
+            
+            
+            if  let label = dict[.label] as? String,
+                let type = dict[.type] as? String,
```

Comment:
func optValue(key: Key) -> Any?
    {
        if(self.hasValue(forKey: key))
        {
            return self[key]!
        }
        else
        {
            return nil
        }
    }
    
    func optString(key : Key) -> String?
    {
        return optValue(key: key) as! String?
    }

--------------------------------------------------------------------------------------------

Anna above is our code for Dictionary Extension . Here we just check if a value exists for the key and we FORCE TYPECAST the value TO STRING TYPE . This might CRASH if the VALUE IS NOT STRING TYPE . That's why I didn't use the extension

Reply:
func optString(key : Key) -> String?
{
    return optValue(key: key) as? String
}
make this changes in our Dictionary extension and use our Dictionary extension for optional values.